satellite/gracefulexit/endpoint.go

```diff
@@ -714,14 +714,24 @@ func (endpoint *Endpoint) updatePointer(ctx context.Context, exitingNodeID storj
 		return nil
 	}
 
-	var toRemove []*pb.RemotePiece
+	pieceMap := make(map[storj.NodeID]*pb.RemotePiece)
 	for _, piece := range remote.GetRemotePieces() {
-		if piece.NodeId == exitingNodeID && piece.PieceNum == pieceNum {
-			toRemove = []*pb.RemotePiece{piece}
-			break
-		}
+		pieceMap[piece.NodeId] = piece
+	}
+
+	var toRemove []*pb.RemotePiece
+	existingPiece := pieceMap[exitingNodeID]
+	if existingPiece != nil && existingPiece.PieceNum == pieceNum {
```

Comment:
We should make sure to return an error in the `else` case. If we don't remove the original piece, we definitely shouldn't add the new one. Or if the piece number of the original piece doesn't match, we shouldn't go any further.